util/topsql/reporter/reporter.go

```diff
@@ -0,0 +1,412 @@
+				MaxDelay:   3 * time.Second,        // Default was 120s.
+			},
+		}),
+	)
+	if err != nil {
+		return nil, nil, err
+	}
+	client := tipb.NewTopSQLAgentClient(conn)
+	return conn, client, nil
+}
+
+// NewRemoteTopSQLReporter creates a new TopSQL struct
+//
+// planBinaryDecoder is a decoding function which will be called asynchronously to decode the plan binary to string
+// MaxStatementsNum is the maximum SQL and plan number, which will restrict the memory usage of the internal LFU cache
+func NewRemoteTopSQLReporter(config *RemoteTopSQLReporterConfig) *RemoteTopSQLReporter {
+	collectTimeout := config.CollectTimeout
+	if collectTimeout > config.CollectInterval || collectTimeout <= 0 {
+		collectTimeout = config.CollectInterval / 2
+	}
+	tsr := &RemoteTopSQLReporter{
+		planBinaryDecoder:  config.PlanBinaryDecoder,
+		MaxStatementsNum:   config.MaxStatementsNum,
+		collectInterval:    config.CollectInterval,
+		collectTimeout:     collectTimeout,
+		topSQLMap:          make(map[string]*topSQLDataPoints),
+		normalizedSQLMap:   make(map[string]string),
+		normalizedPlanMap:  make(map[string]string),
+		collectCPUTimeChan: make(chan *topSQLCPUTimeInput, 1),
+		planRegisterChan:   make(chan *planRegisterJob, 10),
+		agentGRPCAddress:   config.AgentGRPCAddress,
+		instanceID:         config.InstanceID,
+		quit:               make(chan struct{}),
+	}
+
+	go tsr.collectWorker()
+
+	go tsr.registerPlanWorker()
+
+	go tsr.sendToAgentWorker()
+
+	return tsr
+}
+
+// Collect will drop the records when the collect channel is full
+func (tsr *RemoteTopSQLReporter) Collect(timestamp uint64, records []tracecpu.TopSQLCPUTimeRecord) {
+	select {
+	case tsr.collectCPUTimeChan <- &topSQLCPUTimeInput{
+		timestamp: timestamp,
+		records:   records,
+	}:
+	default:
+	}
+}
+
+func (tsr *RemoteTopSQLReporter) collectWorker() {
+	for {
+		input := <-tsr.collectCPUTimeChan
+		tsr.collect(input.timestamp, input.records)
+	}
+}
+
+// collect uses a hashmap to store records in every second, and evict when necessary.
+// This function can be run in parallel with snapshot, so we should protect the map operations with a mutex.
+func (tsr *RemoteTopSQLReporter) collect(timestamp uint64, records []tracecpu.TopSQLCPUTimeRecord) {
+	for _, record := range records {
+		encodedKey := encodeCacheKey(record.SQLDigest, record.PlanDigest)
+		entry, exist := tsr.topSQLMap[string(encodedKey)]
+		if !exist {
+			entry = &topSQLDataPoints{
+				SQLDigest:     record.SQLDigest,
+				PlanDigest:    record.PlanDigest,
+				CPUTimeMsList: []uint32{record.CPUTimeMs},
+				TimestampList: []uint64{timestamp},
+			}
+			tsr.topSQLMap[string(encodedKey)] = entry
+		} else {
+			entry.CPUTimeMsList = append(entry.CPUTimeMsList, record.CPUTimeMs)
+			entry.TimestampList = append(entry.TimestampList, timestamp)
+		}
+		entry.CPUTimeMsTotal += uint64(record.CPUTimeMs)
+	}
+
+	if len(tsr.topSQLMap) <= tsr.MaxStatementsNum {
+		return
+	}
+
+	// find the max CPUTimeMsTotal that should be evicted
```

Comment:
Should find the min CPUTimeMsTotal that should be evicted?

Reply:
It should be "max", imagine this:
We have cpu time ms total of [1, 2, ..., 10], and capacity of 3, then we should find 7 and evict all records less than 7. Here 7 is the "max" of the evicted records.